### ISSUE TITLE
Several models modify caller-provided tensors in place during forward

### ISSUE THREAD

**Issue by Yigtwxx** [huggingface/transformers#47741]
### System Info

- `transformers` version: 5.15.0.dev0 (main @ d79d0c1ed3)
- Platform: Windows-11-10.0.26200-SP0
- Python version: 3.12.0
- PyTorch version (GPU?): 2.13.0+cpu (False)
- Huggingface_hub version: 1.26.0
- Accelerate version: 1.14.0

### Who can help?

@zucchini-nlp @Cyrilvallez

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Several models write into a tensor that the caller passed to `forward`, instead of producing a new
one. The caller still holds a reference to that tensor, so the write escapes the model.

The script below is CPU-only, seeded, and uses randomly initialised tiny configs, so it needs no
downloads:

```python
import torch
from transformers import CanineConfig, CanineForQuestionAnswering, CTRLConfig, CTRLModel

torch.manual_seed(0)

# 1. A model rewrites the caller's label tensor.
canine = CanineForQuestionAnswering(
    CanineConfig(hidden_size=16, num_hidden_layers=1, num_attention_heads=2, intermediate_size=32)
).eval()
start_positions = torch.tensor([9999])
canine(input_ids=torch.tensor([[1, 2, 3, 4]]), start_positions=start_positions, end_positions=torch.tensor([9999]))
print("start_positions after forward:", start_positions.tolist())  # expected [9999]

# 2. A model rewrites the caller's `inputs_embeds`, so the same call twice gives different answers.
ctrl = CTRLModel(CTRLConfig(vocab_size=50, n_embd=16, n_layer=2, n_head=2, dff=32, n_positions=32)).eval()
embeds = ctrl.get_input_embeddings()(torch.tensor([[1, 2, 3, 4]])).detach()
with torch.no_grad():
    first = ctrl(inputs_embeds=embeds).last_hidden_state
    second = ctrl(inputs_embeds=embeds).last_hidden_state
print("same inputs_embeds, two calls agree:", torch.allclose(first, second))
print("max abs difference:", (first - second).abs().max().item())

# 3. The same write makes gradient attribution impossible.
embeds = ctrl.get_input_embeddings()(torch.tensor([[1, 2, 3, 4]])).detach().requires_grad_(True)
try:
    ctrl(inputs_embeds=embeds)
except RuntimeError as exc:
    print("gradient attribution:", exc)
```

Output on `main`:

```
start_positions after forward: [4]
same inputs_embeds, two calls agree: False
max abs difference: 1.0304391384124756
gradient attribution: a leaf Variable that requires grad is being used in an in-place operation.
```

The three symptoms are the same defect seen from three angles:

1. `CanineForQuestionAnswering` calls `start_positions.clamp_(0, ignored_index)`
   ([`modeling_canine.py#L1336`](https://github.com/huggingface/transformers/blob/main/src/transformers/models/canine/modeling_canine.py#L1336)),
   so the caller's label tensor comes back rewritten. Anything computed from that batch after the
   forward pass — a metric, a second loss, a logged example — silently uses the clamped values.
2. `CTRLModel` does `inputs_embeds *= np.sqrt(self.d_model_size)`
   ([`modeling_ctrl.py#L314`](https://github.com/huggingface/transformers/blob/main/src/transformers/models/ctrl/modeling_ctrl.py#L314)),
   so a caller that reuses one embedding tensor across calls gets a different answer every time.
3. The same statement rules out gradient-based attribution (saliency maps, integrated gradients,
   embedding-space perturbation), which builds a leaf `inputs_embeds` with `requires_grad=True` and
   hands it to the model.

### Scope

I checked this by adding a common test that compares each input tensor's version counter across a
`forward` call and running it over every model directory. The version counter rather than the values,
because some of these writes are numerically inert on generated test inputs — `clamp_` on labels that
happen to be in range, a multiply by an all-ones mask — while still bumping the version and breaking
autograd. Sixteen models write into a caller-owned tensor:

| Input | Models |
| --- | --- |
| `inputs_embeds` | blip, ctrl, git, kosmos2, moshi, udop, voxtral_realtime |
| `start_positions` / `end_positions` | canine, luke, markuplm, rembert, splinter |
| `mim_labels` | flava |
| `pixel_values` | tvp |
| `encoder_hidden_states` | moonshine_streaming |
| `decoder_attention_mask` | pix2struct |

Deliberately excluded: modules whose "input" is produced by the model itself rather than the caller —
the MoE expert blocks in mixtral/phimoe/minimax, the SpecAugment masking in the wav2vec2 family, the
DETR-style hybrid encoders whose `inputs_embeds` is the backbone's own feature-map list, and the
`point_embeddings` of the SAM mask decoders. Also excluded are sites guarded by
`if <arg> is None:`, where the tensor was just created locally (csm, florence2, longformer, idefics,
musicgen_melody, ovis2, speecht5, nemotron_asr_streaming).

### Expected behavior

`forward` should treat its arguments as read-only and return new tensors. The rest of the library
already does: 53 question-answering heads (bert, bart, roberta, electra, …) use the out-of-place
`start_positions = start_positions.clamp(0, ignored_index)`, and in `modeling_tvp.py`
`TvpFramePadPrompter` adds its prompt out of place a few lines below a sibling prompter that does not.

Each of the sixteen sites has a one-line out-of-place equivalent that is numerically identical, so the
change is a no-op wherever the tensor was in fact a fresh local. I have the fix and the common test
ready and will open a PR referencing this issue.
